x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java

```diff
@@ -125,10 +138,19 @@ public void execute(IngestDocument ingestDocument, BiConsumer<IngestDocument, Ex
                 }
             });
         }
-        return new InferModelAction.Request(modelId, fields, inferenceConfig);
+        return new InternalInferModelAction.Request(modelId, fields, inferenceConfig, previouslyLicensed);
     }
 
-    void mutateDocument(InferModelAction.Response response, IngestDocument ingestDocument) {
+    void auditWarningAboutLicenseIfNecessary() {
+        if (shouldAudit.get() && shouldAudit.compareAndSet(true, false)) {
```

Comment:
As written this condition defeats the point of using an atomic flag, because another thread can change the value in between the two checks.

I think it should be just `if (shouldAudit.compareAndSet(true, false)) {` because `compareAndSet` will return `false` if the starting value was not equal to the first argument.

Reply:
@droberts195 `.get` is done via a volatile read. So, isn't it WAY cheaper than doing a compareAndSet? I was trying to do a little optimization as this is gonna be a hot path. I can remove it without qualms.